packages/structures/src/Structure.ts

```diff
@@ -0,0 +1,139 @@
+ *
+ * The second layer, in the exported structure is effectively a type cast that allows the getters types to match whatever data template is used
+ *
+ * In order to safely set and access this data, the constructor and patch take data as "partial" and forcibly assigns it to kData. To acommodate this,
+ * kData stores properties as `unknown` when it is omitted, which allows accessing the property in getters even when it may not actually be present.
+ * This is the most technically correct way of representing the value, especially since there is no way to guarantee runtime matches the "type cast."
+ */
+export abstract class Structure<DataType, Omitted extends keyof DataType | '' = ''> {
+	/**
+	 * A construct function used when mixing to allow mixins to set optimized property defaults
+	 *
+	 * @internal
+	 * @remarks This should only be used to set defaults, setting optimized values should be done
+	 * in the mixins `optimizeData` method, which will be called automatically.
+	 * @param data - The full API data received by the Structure
+	 */
+	protected [kMixinConstruct]?(data: Partial<DataType>): void;
+
+	/**
+	 * A function used when mixing to allow mixins to add properties to the result of toJSON
+	 *
+	 * @internal
+	 * @remarks This should only be used to add properties that the mixin optimizes, if the raw
+	 * JSON data is unchanged the property will already be returned.
+	 * @param data - The result of the base class toJSON Structure before it gets returned
+	 */
+	protected [kMixinToJSON]?(data: Partial<DataType>): void;
+
+	/**
+	 * The template used for removing data from the raw data stored for each Structure.
+	 *
+	 * @remarks This template should be overriden in all subclasses to provide more accurate type information.
+	 * The template in the base {@link Structure} class will have no effect on most subclasses for this reason.
+	 */
+	protected static readonly DataTemplate: Record<string, unknown> = {};
+
+	/**
+	 * @returns A cloned version of the data template, ready to create a new data object.
+	 */
+	private getDataTemplate() {
+		return Object.create((this.constructor as typeof Structure).DataTemplate);
+	}
+
+	/**
+	 * The raw data from the API for this struture
+	 *
+	 * @internal
+	 */
+	protected [kData]: Readonly<ReplaceOmittedWithUnknown<Omitted, DataType>>;
+
+	/**
+	 * Creates a new structure to represent API data
+	 *
+	 * @param data - the data from the API that this structure will represent
+	 * @remarks To be made public in subclasses
+	 * @internal
+	 */
+	public constructor(data: Readonly<Partial<DataType>>, ..._rest: unknown[]) {
+		this[kData] = Object.assign(this.getDataTemplate(), data);
+		this[kMixinConstruct]?.(data);
+	}
+
+	/**
+	 * Patches the raw data of this object in place
+	 *
+	 * @param data - the updated data from the API to patch with
+	 * @remarks To be made public in subclasses
+	 * @returns this
+	 * @internal
+	 */
+	protected _patch(data: Readonly<Partial<DataType>>): this {
```

Comment:
So I'll open the wound again. What do we think about making all of these methods (excluding toJSON) symbols? I can esp think `clone()` could at some point be used to represent something else other than the re-create this class?

Reply:
> I can absolutely imagine somebody implementing `clone()` as `send a discord request that recreates this structure with a new id`, like cloning a text channel for example.

You don't have to imagine it as that's precisely what discord.js does. [TextChannel#clone()](<https://discord.js.org/docs/packages/discord.js/14.21.0/TextChannel:Class#clone>)

I think we should rename `clone()` back to `_clone()`, as it also is done in discord.js currently.

Reply:
Since we need to rename it anyways, lets also symbol it, consistency with patch

Reply:
I agree to symbol it in this case